network/firewall.py

```diff
@@ -0,0 +1,396 @@
+import logging
+import json
+from typing import List, Dict
+
+import network.network as nw
+from nftables import Nftables
+
+logger = logging.getLogger(__name__)
+
+nft = Nftables()
+nft.set_json_output(True)
+nft.set_stateless_output(True)
+nft.set_service_output(False)
+nft.set_reversedns_output(False)
+nft.set_numeric_proto_output(True)
+
+
+class Firewall:
+    firewall_has_been_setup = False
+    postrouting_info: Dict[str, str] = {}
+    forwarding_info: Dict[str, str] = {}
+    vm_info: Dict[int, Dict] = {}
+
+    @classmethod
+    def execute_json_nft_commands(cls, commands: List[Dict]) -> int:
+        """Executes a list of nftables commands, and returns the exit status"""
+        commands_dict = {"nftables": commands}
+        try:
+            logger.debug("Validating nftables rules")
+            nft.json_validate(commands_dict)
+        except Exception as e:
+            logger.error(f"Failed to verify nftables rules: {e}")
+
+        logger.debug("Inserting nftables rules")
+        rc, output, error = nft.json_cmd(commands_dict)
+        if rc != 0:
+            logger.error(f"Failed to add nftables rules: {error}")
+
+        return rc
+
+    @classmethod
+    def get_existing_nftables_ruleset(cls) -> Dict:
+        """Retrieves the full nftables ruleset and returns it"""
+        rc, output, error = nft.cmd("list ruleset")
+
+        if rc != 0:
+            logger.error(f"Unable to get nftables ruleset: {error}")
+
+        nft_ruleset = json.loads(output)
+        return nft_ruleset
+
+    @classmethod
+    def get_base_chains_for_hook(cls, hook: str, family: str = "ip") -> List:
+        """Looks through the nftables ruleset and creates a list of
+        all chains that are base chains for the specified hook"""
+        nft_ruleset = cls.get_existing_nftables_ruleset()
+        chains = []
+
+        for entry in nft_ruleset["nftables"]:
```

Comment:
What type are these entries ? This could help in the return type `List[EntryType]`.

Reply:
They're dicts, with potentially other dicts in them, but potentially sometimes it contains lists. And depending on the dict, it might contain another layer of dicts or lists. Furtest I've seen it nesting is about 5 levels deep.